tests/cli/test_remote.py

```diff
@@ -1,7 +1,731 @@
+def test_run_remote(remote: MockRemote):  # noqa: F811
+    command = "echo OK"
+    some_message = "BOBOBOB"
+    remote.expect(
+        host="login.server.mila.quebec",
+        user=unittest.mock.ANY,
+        port=2222,
+        commands=[Command(command, out=some_message.encode())],
+    )
+    r = Remote("mila")
+    result = r.run(command)
+    assert result.stdout == some_message
+
+
+@disable_internet_access
+@pytest.mark.parametrize("keepalive", [0, 123])
+@pytest.mark.parametrize("host", ["mila", "bob"])
+def test_init(
+    keepalive: int,
+    host: str,
+    monkeypatch: pytest.MonkeyPatch,
+):
+    """
+
+    TODO: This test should show exactly what the behaviour of init
+    is so we can isolate it from other tests and avoid having to write versions
+    of other tests based on if the connection was passed or not to the
+    constructor.
+    """
+    import milatools.cli.remote
+
+    # If a class is used as a spec then the return value of the mock (the
+    # instance of the class) will have the same spec. You can use a class as
+    # the spec for an instance object by passing instance=True.
+    # The returned mock will only be callable if instances of the mock are
+    # callable.
+    MockConnection: Mock = create_autospec(Connection, spec_set=True)
+    monkeypatch.setattr(milatools.cli.remote, "Connection", MockConnection)
+    mock_connection: Mock = MockConnection.return_value
+    mock_transport: Mock = create_autospec(
+        paramiko.Transport, spec_set=True, instance=True
+    )
+    mock_connection.transport = mock_transport
+
+    r = Remote(host, keepalive=keepalive)
+
+    MockConnection.assert_called_once_with(host)
+    # The Remote should have created a Connection instance (which happens to be
+    # the mock_connection we already have above.)
+    assert r.connection is mock_connection
+
+    if keepalive:
+        mock_connection.open.assert_called_once()
+        mock_transport.set_keepalive.assert_called_once_with(keepalive)
+    else:
+        mock_connection.open.assert_not_called()
+        mock_transport.set_keepalive.assert_not_called()
+
+
+@pytest.fixture
+def host() -> str:
+    return "mila"
+
+
+@pytest.fixture
+def mock_connection(
+    host: str, mocker: pytest_mock.MockerFixture, internet_enabled: bool
+) -> Connection | Mock:
+    if internet_enabled:
+        # Return a real connection to the host!
+        return Connection(host)
+    MockConnection: Mock = create_autospec(
+        Connection, spec_set=True, _name="MockConnection"
+    )
+    mock_connection: Mock = MockConnection(host)
+    # mock_connection.configure_mock(name="mock_connection")
+    MockTransport: Mock = create_autospec(
+        paramiko.Transport, spec_set=True, _name="MockTransport"
+    )
+    mock_transport: Mock = MockTransport.return_value
+    # mock_transport.configure_mock(name="mock_transport")
+    mock_connection.transport = mock_transport
+    mocker.patch("paramiko.Transport", MockTransport)
```

Comment:
> Wouldn't remote: MockRemote with it's mocked Connection and Transport be usable here? It is necessary to mock ourselves? Note that if this works, this could be merge but I would prefer a comment inciting to look for a solution using what the fabric package already thought for us.

Good question. I'll look back at the code, but I seem to recall that this was necessary (mocking stuff ourselves) because in a lot of cases we pass the output buffer to the functions and this isn't doable with the `MockConnection` which allows for listing expected commands and their inputs/outputs.

> Also, pytests can be executed in parallel and I think this fixture could cause issues if multiples tests modify the same mock_connection fixture? But I might be wrong

Not quite, with pytest-xdist at least each process has all the fixtures it needs for its set of tests (duplicated across processes).

> Also, small comment regarding the var mock_connection I'm a little unsure about shadowing the func of the name name in which it is found but there's no functional issue and this could be merge as is

Yeah good idea. There's no functional issue with this though.